cmd/cache_serve.go

```diff
@@ -46,31 +47,55 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
-func getNSAdsFromDirector() ([]director.NamespaceAd, error) {
+func getNSAdsFromDirector() ([]director.NamespaceAdV2, error) {
 	// Get the endpoint of the director
-	var respNS []director.NamespaceAd
+	var respNS []director.NamespaceAdV2
 	directorEndpoint, err := getDirectorEndpoint()
 	if err != nil {
 		return respNS, errors.Wrapf(err, "Failed to get DirectorURL from config: %v", err)
 	}
 
 	// Create the listNamespaces url
-	directorNSListEndpointURL, err := url.JoinPath(directorEndpoint, "api", "v1.0", "director", "listNamespaces")
+	directorNSListEndpointURL, err := url.JoinPath(directorEndpoint, "api", "v2.0", "director", "listNamespaces")
 	if err != nil {
 		return respNS, err
 	}
 
+	// Attempt to get data from the 2.0 endpoint, if that returns a 404 error, then attempt to get data
+	// from the 1.0 endpoint and convert from V1 to V2
+
 	respData, err := utils.MakeRequest(directorNSListEndpointURL, "GET", nil, nil)
 	if err != nil {
-		if jsonErr := json.Unmarshal(respData, &respNS); jsonErr == nil { // Error creating json
-			return respNS, errors.Wrapf(err, "Failed to make request: %v", err)
+		if strings.Contains(err.Error(), "404") {
+			directorNSListEndpointURL, err = url.JoinPath(directorEndpoint, "api", "v1.0", "director", "listNamespaces")
+			if err != nil {
+				return respNS, err
+			}
+			respData, err = utils.MakeRequest(directorNSListEndpointURL, "GET", nil, nil)
+			var respNSV1 []director.NamespaceAdV1
+			if err != nil {
```

Comment:
If there _is_ an error from `MakeRequest` to `v1.0` endpoint, why do we still want to unmarshal the response data into `respNSV1`? Wouldn't that simply causes an empty `respNSV1` or even json unmarshal error? Or `/v1.0/director/listNamespaces` endpoint _does_ return namespaces even if the response status code != 200?

Reply:
Ah, I see, line 77 is a holdover from old code that I just moved within the if/else statement. It might have been incorrect to begin with. I'll remove it.